Veteran unit production & unlocking through spies.

matija-hustic opened the pull request:
A stab at #7585.

~~Sample veteran unit (E1.upgraded) is defined (using GainsExperience trait, InitialExperience property) and its production made available through tech tree prerequisite (barracks.upgraded under Player.ProvidesPrerequisite (1)) that is normally disabled. The prerequisite gets enabled when a spy infiltrates the appropriate building (BARR or TENT) that has the InfiltrateForUpgrade trait which references the prerequisite trait definition (1) under Player.~~

~~In case the idea is OK, I'd remove the yaml entries and the improvised cameo, leaving just the logic, until the upgraded units' cameos are all available.~~

This PR basically introduces three things.
- Upgrading icons in the production queue.
- Spawning upgraded units.
- Triggering both these upgrades with spies.

GraionDilach commented:
I'd really welcome this trait in OpenRA, however...

I don't think the tag is right. InfiltrateForUpgrade suggests me that the trait implements the upgrade logic while it does not.

Logic dictates it should be named InfiltrateForPrerequisite instead.

matija-hustic committed:
Veteran unit production & unlocking through spies.

GraionDilach commented:
While I can foresee others would find the test-case inappropriate due to #7469, RA2 had 4 stolentech infantry, and it's reimplementation will need such a logic sooner or later.

pchote commented on OpenRA.Mods.Common/Traits/Player/ProvidesPrerequisite.cs:
```diff
@@ -27,6 +27,13 @@ public class ProvidesPrerequisiteInfo : ITraitInfo
 		public readonly bool ResetOnOwnerChange = false;
+
+		[Desc("Does the prerequisite need to be triggered before use?")]
+		public readonly bool NeedsActivation = false;
```
This appears to be duplicating our existing upgrades logic.  You can instead make this implement `UpgradableTrait`, and then override the `UpgradeLevelChanged` method it exposes to replace ActivatePrerequisite.

pchote commented on OpenRA.Mods.RA/Traits/Infiltration/InfiltrateForPrerequisite.cs:
```diff
@@ -0,0 +1,46 @@
+		}
+
+		public void Infiltrated(Actor self, Actor infiltrator)
+		{
```
The implementation of this method doesn't really fit our conventions for trait/actor interactions.  You could instead have an `InfiltrateForActor` trait that spawns an actor that you can then use as a prerequisite directly.  This is how we do support power crates, for example.

matija-hustic replied:
Should I then scrap the InfiltrateForPrerequisite and InfiltrateForActor and just use the InfiltrateForSupportPower, only instead of using the proxy for support power traits, using it for a ProvidesPrerequisite trait?

GraionDilach replied:
I don't like the sound of a support power there - not enough generic. InfiltrateForActor sounds the best option.

matija-hustic replied:
Well, the code will be identical as that of InfiltrateForSupportPower - that already exists. Maybe then rename InfiltrateForSupportPower to InfiltrateForActor?

penev92 replied:
If they do the same thing, then renaming `InfiltrateForSupportPower` to be more generic sounds like the way to go.

penev92 commented:
I can't do a proper review now, but the e1 test case seems wrong to me.
While this is surprisingly clear code-wise, the resulting YAML (when done for 10-15-20-50 actors) would be terrible.
I want to ask reviewers (and yourself) to consider using our upgrade system.
You infiltrate a building, which grants a global upgrade, the appropriate upgrades on the units get triggered (and afaik `GainsExperience` already works with upgrades). Removes the need of defining a second actor.

GraionDilach commented:
I'd say keep the code changes, but abandon the testcase and any upgrade-based infiltration implementation should be separated to another PR.

penev92 commented:
My point here is there might be a better way to implement this. Also going this way would force mods to do some terrible things on YAML level.

GraionDilach commented:
Which I agree with. However the InfiltrateForSupportPower-->InfiltrateForActor rename should still happen, because until this PR, I never thought about using that trait for unlocking new units via infiltration - especially due to the fact that InfiltrateForSupportPower has no trait documentation at all, see https://github.com/OpenRA/OpenRA/wiki/Traits#infiltrateforsupportpower

There are situations when this logic comes handy, but the testcase is clearly not among them.

matija-hustic commented:
I'm not sure I'm following @penev92 - a global upgrade should get granted and upgrades on appropriate units should get triggered? You mean on existing units? Because the idea here is that the upgrade only applies to newly produced units, as in RA2. The existing units are left untouched...

GraionDilach commented:
@matija-hustic , @penev92 has a valid point that your current implementation is flawed due to the modder has to duplicate all affected units to trigger the initial promotion.

If I think about it, maybe the best way to get this effect is to make that InitialExperience trait of yours an Upgradable trait, and upgrade that via the actor gained. This does not give the veterancy mark upon the cameo, true, but a lot less cumbersome for mod maintenance.

pchote commented:
A couple of thoughts:
* `InfiltrateForSupportPower` would be better renamed to `InfiltrateForProxyActor` than `InfiltrateForActor` -- unless somebody wants to massively increase the scope of that trait (which I don't recommend) so that it can spawn regular actors.
* You avoid duplicating actors by having it trigger an upgrade on the production structure, and having the production structure pass an `ExperienceInit` to the actors that it produces.  You then don't need to modify anything on the unit or its traits.

matija-hustic commented:
@pchote, I've just remembered that RA2 goes even further than the production queue. Any paradrops, units left after building destruction and crate-spawned units need to be upgraded as well (if RA2 behavior is to be replicated). This in particular means that the upgrades should take place at actor creation, not when adding to world (paradropped units are not in the world until they land - probably should also be changed for RA2).

My questions - how much sense does it make to use global updates seeing how
1) they are applied when adding to world at the earliest (can be ignored if paradrop will be modified),
2) prerequisites for each such upgrade need to be checked only once on creation and never again, as opposed to other global updates that need continuous checking?

And, should a new trait be better suited, should it
1) check prerequisites on creation (INotifyCreated) using the owning player's tech three (HasPrerequisites) and
2a) calls upgrade manager to grant upgrades definable in the YAML file
or
2b) calls GainsExperience.GiveLevels or GainsExperience.GiveExperience with a certain amount of input from YAMLs?

Phrohdoh commented:
@matija-hustic Correct because there are a lot of things happening here.
Paradropped actors are indeed in the world upon exiting their cargo plane, but they do not reserve space in the `ActorMap`.

`ExperienceInit` is what we will want to use for these cases.

matija-hustic commented:
@Phrohdoh, when I look at the code, right now it seems that the paradropped actors are only added to map once they land (line 78 of Parachute.cs). Plus, they aren't attacked by enemies until they land. RA2 makes it possible to shoot at helpless parachuting units.

Mailaender commented on OpenRA.Mods.RA/Traits/Infiltration/InfiltrateForActor.cs:
```diff
@@ -13,18 +13,18 @@
 namespace OpenRA.Mods.RA.Traits
 {
-	class InfiltrateForSupportPowerInfo : ITraitInfo
+	class InfiltrateForActorInfo : ITraitInfo
```
This needs a OpenRA.Utility mod upgrade rule now.

matija-hustic replied:
Could you clarify, please? I don't understand what you mean. I'm not completely familiar with the OpenRA jargon

penev92 replied:
When traits are changed, mods and maps need to be changed accordingly, otherwise they break. This is done by running the utility commands `--upgrade-mod` and `--upgrade-map`.
Take a look at UpgradeRules.cs, where all "upgrade rules" are located.

matija-hustic replied:
Two things:
1) Can this wait until I get feedback in the PR discussion on how to change the rest of the code? This trait will be renamed to InfiltrateForProxyActor (as @pchote preferred).

2) Running "OpenRA.Utility ra --upgrade-mod" gives
Received args: --upgrade-mod
System.IndexOutOfRangeException: Index was outside the bounds of the array.
   at OpenRA.Mods.Common.UtilityCommands.UpgradeModCommand.Run(ModData modData, String[] args)
   at OpenRA.Utility.Program.Main(String[] args) in d:\Documents\Visual Studio 2013\Projects\OpenRA\OpenRA.Utility\Program.cs:line 54

penev92 replied:
1) Sure. First fix the implementation. Upgrade rules are usually the last thing to be done.
2) No need to run this just yet. Also you can upgrade the official mods and maps by hand if you wish. We'll walk you through it when you get to it.

Mailaender replied:
You could also undo the `InfiltrateForSupportPower` → `InfiltrateForActor` renaming and leave that for another pull request. It doesn't seem a requirement here, but more like a "oh, while you are at it..." which is probably not the best thing to ask new contributors.

GraionDilach replied:
I shall take full responsibility for that, sorry about it.

matija-hustic commented:
Went ahead and done some stuff while waiting for feedback :)

1) I haven't used global upgrades because they are applied only while adding to world, and usually after other traits are notified of adding to world.
2) I haven't implemented levelup through GainsStatUpgrades since this leaves out the rank effect.
3) Making GainsExperience implement UpgradableTrait isn't a perfect solution since silent levelup is sometimes needed, but not always.
4) Implementing icon change through the upgrade system would be more complex than the through the production queue. I am not sure which would be preferred tho.

pchote commented:
I have two issues with this:
* A significant part of this pr now seems to be working around the rank chevron being bogus while paradropping.  This is the fault of the rank code itself (which uses an IEffect instead of the now-standard With* traits), and should be fixed properly (in its own pr) instead of inserting workarounds in N different parts of the code.
* The `UpgradedVariantWatcher` stuff in our core production trait really doesn't sit well with me.  I'm sure there is a more elegant way this could be done.

matija-hustic commented:
I'll get to the `Rank` trait a bit later, but I've moved the icon changing from `ProductionQueue` to a new trait that needs to be attached to the player actor. Is this better?

matija-hustic commented:
Rebased.

penev92 commented on OpenRA.Mods.Common/Traits/GainsExperience.cs:
```diff
@@ -123,6 +127,7 @@ public void ResolveOrder(Actor self, Order order)
 		}
 	}
 
+	// Remove this?
```
Probably not. Why?

matija-hustic replied:
Well it's not used anywhere...

Unit158 replied:
Actually, the experience init is used (and if it isn't, we're either not at those missions yet or we're doing something wrong).

matija-hustic commented:
Well it's not used anywhere...

pchote commented:
@matija-hustic: the *Init classes are used by reflection (they bridge to the actor properties defined in map.yaml), so don't believe your editor saying that it isn't used.

abcdefg30 commented:
Needs a rebase again.

matija-hustic commented:
Rebased.

abcdefg30 commented on OpenRA.Mods.Common/OpenRA.Mods.Common.csproj:
```diff
@@ -687,4 +689,5 @@ cd "$(SolutionDir)"</PostBuildEvent>
   <Target Name="AfterBuild">
   </Target>
   -->
+  <ItemGroup />
```
I think this has to go away?

abcdefg30 commented on OpenRA.Mods.Common/Traits/UpgradesWhenCreated.cs:
```diff
@@ -0,0 +1,44 @@
+			{
+				var ge = self.TraitOrDefault<GainsExperience>();
+				if (ge != null && ge.CanGainLevel)
+					ge.GiveLevels(1, true);
```
Why is the level here hardcoded?

abcdefg30 commented:
Looks good to me and the testcase worked ingame. :+1: / :white_check_mark:

pchote commented on OpenRA.Mods.Common/Traits/GainsExperience.cs:
```diff
@@ -77,32 +77,35 @@ public GainsExperience(ActorInitializer init, GainsExperienceInfo info)
 		{
 			experience += amount;
 
+			var um = self.Trait<UpgradeManager>();
```
Please keep this as `TraitOrDefault` and do the null check.  There is no need to force this trait to require the UpgradeManager.

penev92 replied:
As far as I know, `GainsExperience` doesn't work without upgrades (`UpgradeManager`) at all, so that would make it silently fail.
Declare the dependency explicitly with `Requires<T>`.

matija-hustic replied:
And it should probably be moved to the constructor and cached if I'm not mistaken?

penev92 replied:
Correct.

pchote commented on OpenRA.Mods.Common/Traits/Tooltip.cs:
```diff
@@ -32,10 +32,17 @@ public class TooltipInfo : ITraitInfo, ITooltipInfo
 
 		[Desc("Sequence of the actor that contains the cameo.")]
-		public readonly string Icon = "icon";
+		public string Icon { get; private set; }
```
TraitInfos are supposed to be immutable.  I haven't looked at the code in enough detail to suggest an alternative approach, but please try and find one!

pchote replied:
An alternative solution may be to make `Tooltip` upgradable, and then have the build palette code pick the one that is allowed based on the current upgrades (you'll need new code for that however - the current upgrade checking code only exists for living actors).

pchote commented on OpenRA.Mods.Common/Traits/UpgradesIcons.cs:
```diff
@@ -0,0 +1,61 @@
+	[Desc("Attach this to the player actor. When attached, enables all actors possessing the UpgradesWhenCreated" +
+		"trait to have their production queue icons changed from Tooltip.Icon to Tooltip.IconUpgraded. The icon change" +
+		"occurs when UpgradesWhenCreated.Prerequisites are met.")]
+	public class UpgradesIconsInfo : ITraitInfo
```
Make this implement `Requires<TechTreeInfo>` to avoid initialization-order issues.

matija-hustic commented:
The production palette widgets now interact with the `UpgradesIcons` trait to see if an upgraded icon should be used.

GraionDilach commented:
I was thinking - and you might hate me now - but instead of going with a per-actor cameo switch, why don't you just make an additional layer and put it above the cameo (like how the clock is applied)?

matija-hustic commented:
Well, @GraionDilach, RA2 has separate cameos, as @Phrohdoh linked in #7585

See [RA2 icons](http://xhp.xwis.net/ra2_icons/)

That's why I've done it this way

pchote commented:
That sounds like a good idea to me.  It means that the veterancy-granting code should be nicely contained within the production queue trait and the build palette widget.

GraionDilach commented:
We already did similar changes in RA and the TD mod, such an alteration seems acceptable to me. Not to mention, many of the RA2 units used beta variants of the cameos for upgraded (Crazy Ivan), was missing (Harrier, defences, navy can also gain veterancy in your logic, this was impossible even in Ares-enhanced YR), or went bogus (Cosmonaut to Laser Rocketeer).

This would also let anyone incorporate this trait into any mod without much work.

matija-hustic commented:
The upgraded icons are now implemented overlays.

Phrohdoh commented on OpenRA.Mods.Common/Traits/UpgradesIcons.cs:
```diff
@@ -0,0 +1,137 @@
+		"The icon change occurs when UpgradesWhenCreated.Prerequisites are met.")]
+	public class UpgradesIconsInfo : ITraitInfo, Requires<TechTreeInfo>
+	{
+		[Desc("Image used for this decoration. Defaults to the actor's type.")]
```
This does not default to the actor's type.

matija-hustic replied:
I've done a copy-paste from `WithDecoration` and forgot to adjust the descriptions. I'll fix it in a minute.

Phrohdoh commented on OpenRA.Mods.Common/Traits/UpgradesIcons.cs:
```diff
@@ -0,0 +1,137 @@
+		public readonly ReferencePoints ReferencePoint = ReferencePoints.Top | ReferencePoints.Right;
+
+		[Desc("Pixel offset relative to the actor's bounding box' reference point.")]
+		public readonly int2 Offset = new int2(-7, 6);
```
Where are these values coming from?

matija-hustic replied:
As is usually the case, adjusting for the position that best pleases the eye.

penev92 replied:
Please no.
Such magic default values have caused enough problems in the past. Default values should be the type's default (although we still have a not of exceptions to that rule).

matija-hustic replied:
I've seen it in a lot of other places so I figured it was standard stuff. I'll move it to YAML and make note of this for the future.

penev92 replied:
It used to be a standard thing, it seems. I've been pushing for a while now to get rid of it though (still a long way to go).

Phrohdoh replied:
It won't go away, as stated. Sane defaults for a single mod are better than no default values.

Phrohdoh commented on OpenRA.Mods.Common/Traits/UpgradesIcons.cs:
```diff
@@ -0,0 +1,137 @@
+		public readonly string Palette;
+		public readonly float Scale;
+
+		const string Prefix = "upgradedVariant.";
```
Can this be moved into the Info class?

matija-hustic replied:
Not as far as I know since this is an internally used value and should not be visible outside the code. I know hard-coding is not a desirable practice, but I've just imitated what I've seen in `GlobalUpgradeManager.cs`. 

The problem is, if I remember correctly, that the `TechTree` doesn't associate `Watcher.Key` with the registering `ITechTreeElement`. So in a situation where multiple `ITechTreeElement`s register `Watcher`s with the same `Key`, and one removes its `Watcher`, `Watcher`s for all `ITechTreeElement`s get removed.

Phrohdoh commented on OpenRA.Mods.Common/Traits/UpgradesWhenCreated.cs:
```diff
@@ -0,0 +1,47 @@
+	[Desc("Actors possessing this trait should define the GainsExperience trait beforehand. When the prerequisites are fulfilled, ",
+		"this trait grants a level-up to newly spawned actors. If additionally the actor's owning player defines the CanUpgradeIcons ",
+		"trait, the production queue icon renders with an overlay defined in that trait.")]
+	public class UpgradesWhenCreatedInfo : ITraitInfo
```
I'm not a fan of this name. `Upgrade(s)` indicated that it uses the upgrade system, which this does not.

matija-hustic replied:
How about `LevelupWhenCreated`

Phrohdoh replied:
Sure.

matija-hustic commented:
Fixed as commented.

Phrohdoh commented on OpenRA.Mods.Common/Traits/ProductionIconOverlay.cs:
```diff
@@ -0,0 +1,137 @@
+		public readonly string Palette;
+		public readonly float Scale;
+
+		const string Prefix = "ProductionIconOverlay.";
```
Can this be moved into the Info class?

matija-hustic replied:
Not as far as I know since this is an internally used value and should not be visible outside the code. I know hard-coding is not a desirable practice, but I've just imitated what I've seen in GlobalUpgradeManager.cs.

The problem is, if I remember correctly, that the TechTree doesn't associate Watcher.Key with the registering ITechTreeElement. So in a situation where multiple ITechTreeElements register Watchers with the same Key, and one removes its Watcher, Watchers for all ITechTreeElements get removed.

penev92 replied:
In that case I'd say this calls for a 
`// HACK: ...` with an explanation.
Because looking at the code in this trait, this is pretty redundant for the keys.

Phrohdoh replied:
Just because something isn't desired doesn't mean it is a hack.
We could move this into an info class and write a lint rule for duplicates, if necessary.

matija-hustic replied:
But doesn't a thing like this require a proper solution in `TechTree`? If not for this trait, then for the `GlobalUpgradeManager`

WolfGaming replied:
This could be fixed if the tech tree would handle making the keys and just returns them when a TechTreeElement is added, I tried fixing that but ran out of time, might have a look again.

Meanwhile I guess this "hack" is ok.

penev92 replied:
Undesired workarounds to make code that doesn't work work are hacks. The proper fix is indeed in TechTree, but in the mean time, that comment will note that something fixing.

penev92 commented on OpenRA.Mods.Common/Traits/ProductionIconOverlay.cs:
```diff
@@ -0,0 +1,137 @@
+		public readonly int2 Offset = int2.Zero;
+
+		[Desc("Visual scale of the overlay.")]
+		public readonly float Scale = 2f;
```
I'd say this is another magic default value that should go away.
Overwriting the default (2) when you want 1 is strange at best.

Phrohdoh commented on mods/ra/rules/misc.yaml:
```diff
@@ -356,6 +356,10 @@ powerproxy.paratroopers:
 		DisplayBeacon: true
 		BeaconPoster: pinficon
 
+barracks.upgraded
```
Missing `:`

penev92 commented on OpenRA.Mods.Common/Traits/LevelupWhenCreated.cs:
```diff
@@ -0,0 +1,47 @@
+
+namespace OpenRA.Mods.Common.Traits
+{
+	[Desc("Actors possessing this trait should define the GainsExperience trait beforehand. When the prerequisites are fulfilled, ",
```
This is not consistent with the code below, which cautiously checks if the actor has `GainsExperience`.
Also "actor should define trait X" is handled by `Requires<>` (which you are missing), and "beforehand" is handled by our actor initializer, which initializes the traits in order (required traits get initialized first).
Lastly you are doing the trait lookup in `AddedToWorld()`, not in the constructor, so at that point all traits are initialized (meaning you don't care about the order).

To sum up: Add `Requires<>`, change `.TraitOrDefault()` to `.Trait()` and skip the null check.

matija-hustic replied:
@penev92, isn't `Requires<Trait>` only needed when `Trait` is used in the constructor?

I'll remove the "beforehand" from the comment and add change `.TraidOrDefault()` to `.Trait()`, but give me a confirmation for the `Requires<>` pls.

matija-hustic replied:
Answered by @pchote in IRC.

penev92 commented on OpenRA.Mods.Common/Traits/LevelupWhenCreated.cs:
```diff
@@ -0,0 +1,47 @@
+
+		public void AddedToWorld(Actor self)
+		{
+			if (self.Owner.PlayerActor.Trait<TechTree>().HasPrerequisites(info.Prerequisites))
```
Please invert the condition here to reduce nesting (ReSharper likes to suggest that...).

WolfGaming replied:
This will fail with a null pointer exception when no prerequisites are defined, please change
```
public readonly string[] Prerequisites = null;
```
in LevelupWhenCreatedInfo to
```
public readonly string[] Prerequisites = { };
```

Mailaender commented on OpenRA.Mods.Common/Traits/ProductionIconOverlay.cs:
```diff
@@ -0,0 +1,142 @@
+		public readonly string Image = "rank";
+
+		[Desc("Sequence used for the overlay (cannot be animated).")]
+		public readonly string Sequence = "rank";
```
Please add `[SequenceReference("Image")]` here. See https://github.com/OpenRA/OpenRA/issues/7709.

Mailaender commented on mods/ra/rules/player.yaml:
```diff
@@ -65,4 +65,7 @@ Player:
 	GlobalUpgradeManager:
 	EnemyWatcher:
+	ProductionIconOverlay:
+		Offset: -7, 6
```
You can tweak this a bit for additional polishing:

![image](https://cloud.githubusercontent.com/assets/756669/8392337/91c6c1a0-1ce7-11e5-945a-d434426d6869.png)

GraionDilach replied:
I'd say take @tomson26's chevron from http://www.ppmforums.com/viewtopic.php?p=508926#508926 and place it to the top left corner.

Mailaender commented:
Works as promised. :white_check_mark: Almost done here.

matija-hustic commented:
Fixed as commented. Rebased.

@GraionDilach, may I ask for your help with the cameo overlay please? It doesn't scale nicely and I'm not sure I've converted it correctly.

GraionDilach commented:
Okay, I'll do.

GraionDilach commented on OpenRA.Mods.Common/Traits/ProductionIconOverlay.cs:
```diff
@@ -0,0 +1,142 @@
+	public class ProductionIconOverlayInfo : ITraitInfo, Requires<TechTreeInfo>
+	{
+		[Desc("Image used for the overlay.")]
+		public readonly string Image = "upgico";
```
These defaullt values sound horrid and maybe shouldn't even exist.

GraionDilach commented on mods/ra/rules/misc.yaml:
```diff
@@ -355,6 +355,10 @@ powerproxy.paratroopers:
 		BeaconPoster: pinficon
 
+barracks.upgraded:
+	AlwaysVisible:
```
Why is this?

matija-hustic replied:
#8477

GraionDilach replied:
Ah, yeah, you're right, sorry.

GraionDilach commented:
And I redid the chevron.

Mailaender commented:
You may also want to disable this in mission Allies05 to avoid frustration due to an unaccomplishable mission.

GraionDilach commented:
Pinging @ScottNZ to review how this affects/what changes are needed in terms of overall gameplay.

The code looks okay now.

GraionDilach commented on OpenRA.Mods.Common/Traits/ProductionIconOverlay.cs:
```diff
@@ -0,0 +1,142 @@
+		[SequenceReference("Image")] public readonly string Sequence = null;
+
+		[Desc("Palette to render the sprite in. Reference the world actor's PaletteFrom* traits.")]
+		public readonly string Palette = "effect";
```
I'd say this should default to chrome, to be honest.

matija-hustic commented:
@Mailaender, I never asked if this was to be used in RA1 at all. I'll need the final form for YAMLs before this is good to go, and then I'll also disable in Allies05 whatever actually gets enabled with this PR.

matija-hustic commented:
@GraionDilach, thank you for the help with the cameos.

GraionDilach commented:
I thank you that you planned this out with custom palette support - that's the sole reason it can look such nice.

GraionDilach commented:
A random idea: the [TS Armory](http://cnc.wikia.com/wiki/Civilian_Armory) could be a usecase, by capturing the building, instead of infilitrating it.

Mailaender commented:
> I never asked if this was to be used in RA1 at all.

Why are you polishing the cameo and committing it then if you don't want this in at all?

matija-hustic commented:
@Mailaender, RA2 needs this...

Mailaender commented:
Why not just keep it for the RA mod, too? Just the chevron needs to be a bit smaller to not overlap with the soldiers head.

matija-hustic commented:
I was just asking about whether this is wanted in RA1, since the RA2 style paratroopers in RA1 for example are kind of controversial - they probably require rebalancing and not all the players of RA1 might like having the RA2 functionality in RA1.
But sure, I'll add the YAMLs for upgradable infantry, vehicles and planes. I'll probably see in IRC about certain units in particular as well as ships, since infiltrating the sub pens and naval yards might be impossible...

GraionDilach commented:
I still say that the TS armory should grant a level to all infantry when captured. That can be a pretty good usecase. Meanwhile we can open a topic on Sliepnirstuff to generate player feedback upon this feature.

Also I think that overlapping is acceptable - although since my chevron uses Additive blending, probably the downscaling to 0.5/0.75 still looks good as well.

matija-hustic commented:
@GraionDilach, I haven't played TS that much so I'm not familiar with the armory. What you're saying is that it should grant a level up to all existing infantry? Or only newly spawned infantry? Should it render the production icon overlay?

Mailaender commented:
Imho the spy in our RA mod is still pretty useless and all the ideas on this tracker like https://github.com/OpenRA/OpenRA/issues/2901 and https://github.com/OpenRA/OpenRA/issues/5298 are pretty crap so I vote for keeping this. We have several RA2/TS features such as bridge repair/C4, resource explosions in the classic mods which make them better.

GraionDilach commented:
@matija-hustic The Armory in TS was useless, pretty much a joke. You sent in 5 infantry, they came out on with max experience and the building became useless thereafter. Not many mods reused that feat.

What you practically developed here, is pretty much the [Ares Academies](http://ares-developers.github.io/Ares-docs/new/academy.html) feature, which the Mental Omega mod plans to reuse for their Armory - when captured, all newly trained infantry starts with a free level.

That's a great use-case and makes the building worthwhile.

@Mailaender, wow, you're right, those ideas are indeed terrible. In that scenario, I agree, let's make the Spy infiltrating barracks to grant a free level for infantry, war factory to grant for vehicles and either helipad either airfield to grant veterancy for flying things. Maybe even the Hijacker should be allowed this to be a counterpart - it can steal vehicles instead of disguising.

penev92 commented on OpenRA.Mods.Common/Traits/LevelupWhenCreated.cs:
```diff
@@ -0,0 +1,49 @@
+namespace OpenRA.Mods.Common.Traits
+{
+	[Desc("Actors possessing this trait should define the GainsExperience trait. When the prerequisites are fulfilled, ",
+		"this trait grants a level-up to newly spawned actors. If additionally the actor's owning player defines the CanUpgradeIcons ",
```
I believe the part about `CanUpgradeIcons` is outdated? That is now `ProductionIconOverlay`, I think.

penev92 commented:
As discussed on IRC today, this has my +1 once more units get the level up bonus in RA.

matija-hustic committed:
RA Upgrade YAMLs.

matija-hustic committed:
TS Upgrade YAMLs.

matija-hustic commented:
OK, so I've added YAMLs for a bunch of other units in RA + that Armory addition for TS. I've left
* the code changes in the first commit,
* RA additions in the second commit and
* TS additions in the third commit.

You're probably aware of this, but it's kinda weird that there exist units that implement `GainsStatUpgrades` which are either uncapable of killing, or are unarmed but able to kill (vehicles crushing). In both of these cases, firepower, reload and accuracy upgrades should probably not be granted once #8572 is merged. In the first case, maybe everybody inheriting `^GainsExperience` should be reconsidered?

I'll need help deciding who gets `LevelupWhenCreated` and who doesn't.

I have not put a `LevelupWhenCreated` on any units without armaments. The only possible reason for this to be reconsidered is upgrading speed and armor for unarmed units able to kill (again e.g. vehicles crushing).

I haven't put a `LevelupWhenCreated` on the following units in RA:
* Infantry: Attack Dog, Spy, ^ArmedCivilian (Agent Delphi, General), Zombie, Ant
* Infantry*: Medic, Mechanic
* Vehicles: Minelayer, Demolition Truck, MAD Tank
* Aircraft: BADR.Bomber

and in TS:
* Medic, Chem Spray Infantry, Umagon, Mutant, Mutant Soldier, Mutant Sergeant, Tiberian Fiend, Adult Visceroid, CIV1, CIV3

GraionDilach commented:
Umagon, Mutant, Mutant Soldier, Mutant Sergeant should get it. IIRC the Chem Spray infantry was unused in TS entirely. Umagion is even a GDI heroine in Firestorm, not mission-only á lá Elite cadres.

:+1: otherwise.

pchote commented on OpenRA.Mods.Common/Traits/GainsExperience.cs:
```diff
@@ -77,18 +78,20 @@ public GainsExperience(ActorInitializer init, GainsExperienceInfo info)
 
 			if (init.Contains<ExperienceInit>())
-				GiveExperience(init.Get<ExperienceInit, int>());
+				GiveExperience(init.Get<ExperienceInit, int>(), true);
```
Maybe extract this to an Info parameter?

matija-hustic replied:
The silencing bool?

pchote replied:
Yeah

matija-hustic replied:
How's `SuppressLevelupAnimation` for a name?

pchote replied:
Sounds good to me.  Don't forget to add it to your new trait, too.

pchote commented on OpenRA.Mods.Common/Traits/LevelupWhenCreated.cs:
```diff
@@ -0,0 +1,49 @@
+			this.info = info;
+		}
+
+		public void AddedToWorld(Actor self)
```
This may be called multiple times over the lifetime of an actor (e.g. when exiting a transport).  Free levels for everybody!

You can use INotifyCreated instead.

pchote commented on OpenRA.Mods.Common/Traits/LevelupWhenCreated.cs:
```diff
@@ -0,0 +1,49 @@
+		public object Create(ActorInitializer init) { return new LevelupWhenCreated(init, this); }
+	}
+
+	public class LevelupWhenCreated : INotifyAddedToWorld
```
I'm not convinced about this trait name.  How about something like `ProduceableAsVeteran` to make it clearer that this is tied with ProductionIconOverlay?

penev92 replied:
`ProduceableAs<something_else>`, because you can specify an arbitrary initial level count.

matija-hustic replied:
`ProduceableWithLevelup` then? As always, I'm open to suggestions

pchote replied:
`ProduceableWithLevel`?

pchote commented on OpenRA.Mods.Common/Traits/ProductionIconOverlay.cs:
```diff
@@ -0,0 +1,142 @@
+		public object Create(ActorInitializer init) { return new ProductionIconOverlay(init, this); }
+	}
+
+	public class ProductionIconOverlay : ITechTreeElement
```
This name is a bit too generic for what it does.  How about calling it something like `VeteranProductionIconOverlay`?

pchote replied:
Maybe define an `IProductionIconOverlay` interface to decouple this from the production / observer widgets.

Mailaender commented on mods/ts/rules/civilian-structures.yaml:
```diff
@@ -641,6 +641,8 @@ CAARMR:
 	RenderBuilding:
 		Palette: player
+	ProvidesPrerequisite:
+		barracks.upgraded
```
```yaml
		Prerequisite: barracks.upgraded
```

Also needs a

```yaml
	Capturable:
```

GraionDilach replied:
It doesn't, this actor is off-map.

matija-hustic commented:
Fixed.

matija-hustic commented:
OK, so now it oughta be fine.

Mailaender commented:
Works as promised. :white_check_mark:

Mailaender commented:
[Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/ec4cee53d9d35d50aad40cf901e1f54bcf2dba53...dee280bc26c0ea8fe6666da5a3dd8f817a2620a0)

Mailaender commented:
Thanks a a lot. This makes spies a lot more useful in-game.

Mailaender commented:
You should add yourself to https://github.com/OpenRA/OpenRA/blob/bleed/AUTHORS.

GraionDilach commented:
https://github.com/OpenRA/OpenRA/blob/bleed/AUTHORS#L76
